Bug: disabling default-features on `webpki` fails builds that depend on quinn on Linux

Apparently, the fact that `rustls-platform-verifier` disables default features on its `webpki`-dependency declaration causes trouble for downstream packages when building on Linux.

For example, a simple package with the following dependencies will fail to compile with the error below:

```
[dependencies]
quinn = "0.11.6"
# rustls-webpki = { version = "0.102", features = ["std"] }
```

build error:

```
error[E0277]: the trait bound `webpki::Error: std::error::Error` is not satisfied
   --> /<snap>/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rustls-platform-verifier-0.4.0/src/verification/others.rs:234:40
    |
234 |             other_err.0.downcast_ref::<webpki::Error>()
    |                         ------------   ^^^^^^^^^^^^^ the trait `std::error::Error` is not implemented for `webpki::Error`
    |                         |
    |                         required by a bound introduced by this call
    |
note: required by a bound in `<(dyn std::error::Error + Send + Sync + 'static)>::downcast_ref`
   --> /<snap>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/error.rs:271:28
    |
271 |     pub fn downcast_ref<T: Error + 'static>(&self) -> Option<&T> {
    |                            ^^^^^ required by this bound in `<dyn Error + Send + Sync>::downcast_ref`

```

Uncommenting the `rustls-webpki` dependency above and explicitly enabling `std` fixes the issue.

### System and cargo version

```
$ cargo --version
cargo 1.85.0 (d73d2caf9 2024-12-31)

$ uname -a 
uname -a
Linux dsd-machine 6.8.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Sep 11 15:25:05 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

--- a/rustls-platform-verifier/Cargo.toml
+++ b/rustls-platform-verifier/Cargo.toml
@@ -29,7 +29,7 @@ cert-logging = ["base64"]
 docsrs = ["jni"]
 
 [dependencies]
-rustls = { version = "0.23.24", default-features = false, features = ["std"] }
+rustls = { version = "0.23.25", default-features = false, features = ["std"] }
 log = { version = "0.4" }
 base64 = { version = "0.22", optional = true } # Only used when the `cert-logging` feature is enabled.
 jni = { version = "0.21", default-features = false, optional = true } # Only used during doc generation

--- a/rustls-platform-verifier/Cargo.toml
+++ b/rustls-platform-verifier/Cargo.toml
@@ -43,7 +43,7 @@ webpki = { package = "rustls-webpki", version = "0.102", default-features = fals
 rustls-platform-verifier-android = { path = "../android-release-support", version = "0.1.0" }
 jni = { version = "0.21", default-features = false }
 webpki = { package = "rustls-webpki", version = "0.102", default-features = false }
-android_logger = { version = "0.13", optional = true } # Only used during testing.
+android_logger = { version = "0.15", optional = true } # Only used during testing.
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
 webpki = { package = "rustls-webpki", version = "0.102", default-features = false }

--- a/rustls-platform-verifier/Cargo.toml
+++ b/rustls-platform-verifier/Cargo.toml
@@ -29,7 +29,7 @@ cert-logging = ["base64"]
 docsrs = ["jni"]
 
 [dependencies]
-rustls = { version = "0.23.16", default-features = false, features = ["std"] }
+rustls = { version = "0.23.24", default-features = false, features = ["std"] }
 log = { version = "0.4" }
 base64 = { version = "0.22", optional = true } # Only used when the `cert-logging` feature is enabled.
 jni = { version = "0.21", default-features = false, optional = true } # Only used during doc generation
@@ -37,16 +37,16 @@ once_cell = "1.9"
 
 [target.'cfg(all(unix, not(target_os = "android"), not(target_vendor = "apple"), not(target_arch = "wasm32")))'.dependencies]
 rustls-native-certs = "0.8"
-webpki = { package = "rustls-webpki", version = "0.102", default-features = false }
+webpki = { package = "rustls-webpki", version = "0.103", default-features = false }
 
 [target.'cfg(target_os = "android")'.dependencies]
 rustls-platform-verifier-android = { path = "../android-release-support", version = "0.1.0" }
 jni = { version = "0.21", default-features = false }
-webpki = { package = "rustls-webpki", version = "0.102", default-features = false }
+webpki = { package = "rustls-webpki", version = "0.103", default-features = false }
 android_logger = { version = "0.15", optional = true } # Only used during testing.
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
-webpki = { package = "rustls-webpki", version = "0.102", default-features = false }
+webpki = { package = "rustls-webpki", version = "0.103", default-features = false }
 webpki-root-certs = "0.26"
 
 # BSD targets require webpki-roots-certs for the real-world verification tests.

--- a/rustls-platform-verifier/Cargo.toml
+++ b/rustls-platform-verifier/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rustls-platform-verifier"
-version = "0.5.0"
+version = "0.5.1"
 authors = ["ComplexSpaces <complexspacescode@gmail.com>", "1Password"]
 description = "rustls-platform-verifier supports verifying TLS certificates in rustls with the operating system verifier"
 keywords = ["tls", "certificate", "verification", "os", "native"]

--- a/rustls-platform-verifier/src/tests/verification_mock/mod.rs
+++ b/rustls-platform-verifier/src/tests/verification_mock/mod.rs
@@ -21,15 +21,21 @@
     not(target_os = "visionos")
 ))]
 
+use std::convert::TryFrom;
+use std::net::IpAddr;
+#[cfg(not(any(target_vendor = "apple", windows)))]
+use std::net::{Ipv4Addr, Ipv6Addr};
+use std::sync::Arc;
+
+use rustls::client::danger::ServerCertVerifier;
+use rustls::pki_types;
+#[cfg(not(any(target_vendor = "apple", windows)))]
+use rustls::pki_types::{DnsName, ServerName};
+use rustls::{CertificateError, Error as TlsError, OtherError};
+
 use super::TestCase;
 use crate::tests::{assert_cert_error_eq, ensure_global_state, verification_time};
 use crate::verification::{EkuError, Verifier};
-use rustls::client::danger::ServerCertVerifier;
-use rustls::pki_types;
-use rustls::{CertificateError, Error as TlsError, OtherError};
-use std::convert::TryFrom;
-use std::net::IpAddr;
-use std::sync::Arc;
 
 macro_rules! mock_root_test_cases {
     { $( $name:ident [ $target:meta ] => $test_case:expr ),+ , } => {
@@ -258,6 +264,12 @@ mock_root_test_cases! {
         chain: &[ROOT1_INT1_EXAMPLE_COM_GOOD, ROOT1_INT1],
         stapled_ocsp: None,
         verification_time: verification_time(),
+        #[cfg(not(any(target_vendor = "apple", windows)))]
+        expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForNameContext {
+            expected: ServerName::DnsName(DnsName::try_from("example.org").unwrap()),
+            presented: vec!["DnsName(\"example.com\")".to_owned()]
+        })),
+        #[cfg(any(target_vendor = "apple", windows))]
         expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForName)),
         other_error: no_error!(),
     },
@@ -266,6 +278,12 @@ mock_root_test_cases! {
         chain: &[ROOT1_INT1_LOCALHOST_IPV4_GOOD, ROOT1_INT1],
         stapled_ocsp: None,
         verification_time: verification_time(),
+        #[cfg(not(any(target_vendor = "apple", windows)))]
+        expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForNameContext {
+            expected: ServerName::IpAddress(pki_types::IpAddr::V4(Ipv4Addr::from([198, 168, 0, 1]).into())),
+            presented: vec!["IpAddress(127.0.0.1)".to_owned()],
+        })),
+        #[cfg(any(target_vendor = "apple", windows))]
         expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForName)),
         other_error: no_error!(),
     },
@@ -274,6 +292,12 @@ mock_root_test_cases! {
         chain: &[ROOT1_INT1_LOCALHOST_IPV6_GOOD, ROOT1_INT1],
         stapled_ocsp: None,
         verification_time: verification_time(),
+        #[cfg(not(any(target_vendor = "apple", windows)))]
+        expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForNameContext {
+            expected: ServerName::IpAddress(pki_types::IpAddr::V6(Ipv6Addr::from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 198, 168, 0, 1]).into())),
+            presented: vec!["IpAddress(0::1)".to_owned()],
+        })),
+        #[cfg(any(target_vendor = "apple", windows))]
         expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForName)),
         other_error: no_error!(),
     },

--- a/rustls-platform-verifier/src/tests/verification_real_world/mod.rs
+++ b/rustls-platform-verifier/src/tests/verification_real_world/mod.rs
@@ -34,13 +34,16 @@
 //! they will continue roughly as though they received a "Good" response.
 //! Thus we don't expect these tests to be flaky w.r.t. that, except for
 //! potentially poor performance.
+
+use rustls::client::danger::ServerCertVerifier;
+use rustls::pki_types;
+#[cfg(not(any(target_vendor = "apple", windows)))]
+use rustls::pki_types::{DnsName, ServerName};
+use rustls::{CertificateError, Error as TlsError};
+
 use super::TestCase;
 use crate::tests::{assert_cert_error_eq, ensure_global_state, verification_time};
 use crate::Verifier;
-use rustls::client::danger::ServerCertVerifier;
-use rustls::pki_types;
-use rustls::{CertificateError, Error as TlsError};
-use std::convert::TryFrom;
 
 // This is the certificate chain presented by one server for
 // my.1password.com when this test was updated 2023-08-01. It is
@@ -200,6 +203,12 @@ real_world_test_cases! {
         chain: VALID_1PASSWORD_COM_CHAIN,
         stapled_ocsp: None,
         verification_time: verification_time(),
+        #[cfg(not(any(target_vendor = "apple", windows)))]
+        expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForNameContext {
+            expected: ServerName::DnsName(DnsName::try_from("agilebits.com").unwrap()),
+            presented: vec!["DnsName(\"*.1password.com\")".to_owned(), "DnsName(\"1password.com\")".to_owned()],
+        })),
+        #[cfg(any(target_vendor = "apple", windows))]
         expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForName)),
         other_error: no_error!(),
     },
@@ -220,6 +229,12 @@ real_world_test_cases! {
         chain: VALID_UNRELATED_CHAIN,
         stapled_ocsp: None,
         verification_time: verification_time(),
+        #[cfg(not(any(target_vendor = "apple", windows)))]
+        expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForNameContext {
+            expected: ServerName::DnsName(DnsName::try_from("my.1password.com").unwrap()),
+            presented: vec!["DnsName(\"agilebits.com\")".to_owned(), "DnsName(\"www.agilebits.com\")".to_owned()],
+        })),
+        #[cfg(any(target_vendor = "apple", windows))]
         expected_result: Err(TlsError::InvalidCertificate(CertificateError::NotValidForName)),
         other_error: no_error!(),
     },

--- a/rustls-platform-verifier/src/verification/others.rs
+++ b/rustls-platform-verifier/src/verification/others.rs
@@ -222,14 +222,10 @@ impl Default for Verifier {
 }
 
 fn map_webpki_errors(err: TlsError) -> TlsError {
-    if let TlsError::InvalidCertificate(CertificateError::Other(other_err)) = &err {
-        if let Some(webpki::Error::RequiredEkuNotFound) =
-            other_err.0.downcast_ref::<webpki::Error>()
-        {
-            return TlsError::InvalidCertificate(CertificateError::Other(OtherError(Arc::new(
-                super::EkuError,
-            ))));
-        }
+    if let TlsError::InvalidCertificate(CertificateError::InvalidPurpose) = &err {
+        return TlsError::InvalidCertificate(CertificateError::Other(OtherError(Arc::new(
+            super::EkuError,
+        ))));
     }
 
     err

